packages/pubsub/src/connection-pool.js

```diff
@@ -0,0 +1,245 @@
+/*!
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*!
+ * @module pubsub/connectionPool
+ */
+
+'use strict';
+
+var arrify = require('arrify');
+var common = require('@google-cloud/common');
+var each = require('async-each');
+var events = require('events');
+var extend = require('extend');
+var is = require('is');
+var util = require('util');
+var uuid = require('uuid');
+
+/**
+ * ConnectionPool is used to manage the stream connections created via
+ * StreamingPull rpc.
+ *
+ * @param {module:pubsub/subscription} subscription - The subscription to create
+ *     connections for.
+ * @param {object=} options - Pool options.
+ * @param {number} options.maxConnections - Number of connections to create.
+ *     Default: 5.
+ * @param {number} options.ackDeadline - The ack deadline to send when
+ *     creating a connection.
+ */
+function ConnectionPool(subscription, options) {
+  this.subscription = subscription;
+  this.connections = new Map();
+  this.isPaused = false;
+  this.isOpen = false;
+
+  this.settings = extend({
+    maxConnections: 5,
+    ackDeadline: 10000
+  }, options);
+
+  events.EventEmitter.call(this);
+
+  this.open();
+}
+
+util.inherits(ConnectionPool, events.EventEmitter);
+
+/**
+ * Acquires a connection from the pool. Optionally you can specify an id for a
+ * specific connection, but if it is no longer available it will return the
+ * first available connection.
+ *
+ * @param {string=} id - The id of the connection to retrieve.
+ * @param {function} callback - The callback function.
+ * @param {?error} callback.err - An error returned while acquiring a
+ *     connection.
+ * @param {stream} callback.connection - A duplex stream.
+ */
+ConnectionPool.prototype.acquire = function(id, callback) {
+  var self = this;
+
+  if (is.fn(id)) {
+    callback = id;
+    id = null;
+  }
+
+  if (!this.isOpen) {
+    callback(new Error('Connection pool is closed.'));
```

Comment:
Truthfully I don't think this condition will ever occur, creating a connection pool opens it immediately and when a subscription is closed we close the pool and discard it immediately. If that error were to pop up, I would think it was because of a user manually closed the pool.